src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -4428,6 +4429,55 @@ class SymbolTableVisitor : public CommonVisitor<SymbolTableVisitor> {
         //Implemented in BodyVisitor
     }
 
+    void import_cpython(const AST::FunctionDef_t &x, SymbolTable *parent_scope) {
+        std::vector<std::string> pybind_funcs = {
+            "Py_Initialize",
+            "Py_IsInitialized", 
+            "PyRun_SimpleString", 
+            "Py_DecodeLocale", 
+            "PySys_SetArgv", 
+            "Py_FinalizeEx",
+            "PyUnicode_FromString",
+            "PyUnicode_AsUTF8AndSize",
+            "PyImport_Import",
+            "Py_DecRef",
+            "Py_IncRef",
+            "PyObject_GetAttrString",
+            "PyTuple_New",
+            "PyTuple_SetItem",
+            "PyObject_CallObject",
+            "PyLong_AsLongLong",
+            "PyLong_AsUnsignedLongLong",
+            "PyLong_FromLongLong",
+            "PyLong_FromUnsignedLongLong",
+            "PyFloat_FromDouble",
+            "PyFloat_AsDouble",
+            "PyBool_FromLong",
+            "PyObject_IsTrue",
+        };
+        Str s;
+        AST::alias_t *module_symbols =
+            al.allocate<AST::alias_t>(pybind_funcs.size());
+        
+        for (size_t i = 0; i < pybind_funcs.size(); i++) {
+            s.from_str(al, pybind_funcs.at(i));
+            (module_symbols + i)->loc = x.base.base.loc;
+            (module_symbols + i)->m_name = s.c_str(al);
+            (module_symbols + i)->m_asname = nullptr;
+        }
+
+        AST::ImportFrom_t *imports =
+            (AST::ImportFrom_t*)AST::make_ImportFrom_t(
+                al, x.base.base.loc,
+                (char*)"cpython_bindings", module_symbols,
+                pybind_funcs.size(), 0);
+        
+        SymbolTable *function_scope = current_scope;
+        current_scope = parent_scope;
+        visit_ImportFrom(*imports);
+        current_scope = function_scope;
+    }
```

Comment:
Why are these needed in AST->ASR?

Ideally the initial ASR has nothing to do with Python C/API. Then the ASR->ASR pass translates `bind(python)` into C/API calls in ASR.

Reply:
Exactly, even I was confused in my last meet with Vipul. Looking into the changes, I think I understand what he is doing here.

He has a new runtime module `cpython_bindings.py` which contains the declarations to the `C` functions in `Python.h` that we need. In the above change, he is just creating `AST` code for the following:

```
from cpython_bindings import Py_Initialize, Py_IsInitialized, PyRun_SimpleString, Py_DecodeLocale, PySys_SetArgv, Py_FinalizeEx, PyUnicode_FromString, PyUnicode_AsUTF8AndSize, PyImport_Import, Py_DecRef, Py_IncRef, PyObject_GetAttrString, PyTuple_New, PyTuple_SetItem, PyObject_CallObject, PyLong_AsLongLong, PyLong_AsUnsignedLongLong, PyLong_FromLongLong, PyLong_FromUnsignedLongLong, PyFloat_FromDouble, PyFloat_AsDouble, PyBool_FromLong, PyObject_IsTrue
```

When he visits the `AST` code, `lpython` reads/loads the `runtime/cpython_bindings.py` and automatically generates the declarations for the above `BindC` functions.

@Vipul-Cariappa  Initially this approach sounded interesting to me, but now after going through the changes, it seems like we should instead write the ASR code for these directly in the `python_bind` pass. I think if we simply extend the current `pybind_funcs` array with the parameter types and return type info (for example specifying types like here https://github.com/lcompilers/lpython/blob/9374feb17362d8046db1e4060d633f7ff9887100/src/libasr/codegen/asr_to_wasm.cpp#L206), we should be able to run a loop and generate the ASR nodes for declaring these. Then, we ideally would not be making any changes in the AST->ASR level. And we can also remove the `runtime/cpython_bindings.py` file.

Reply:
Ok. I will update the code to generate the required function declarations in the `python_bind` pass.

Reply:
> Initially this approach sounded interesting to me, but now after going through the changes, it seems like we should instead write the ASR code for these directly in the `python_bind` pass.

Yes, exactly.